libs/@local/hql/span/src/lib.rs

```diff
@@ -0,0 +1,71 @@
+#![feature(never_type)]
+extern crate alloc;
+
+pub mod storage;
+pub mod tree;
+
+pub use text_size::{TextRange, TextSize};
+
+/// Represents a unique identifier for a span of text within a source file.
+///
+/// This struct serves as an opaque unique identifier that can be used to reference a specific
+/// span, which needs to be looked up in the storage system.
+///
+/// ## Compactness
+///
+/// `SpanId` is designed to be as small as possible. It is only 4 bytes in size, allowing it
+/// to be stored compactly and passed efficiently.
+///
+/// ## Usage
+///
+/// Since `SpanId` is just an identifier, the actual span information must be retrieved
+/// from the storage system [`SpanStorage`]. This differs from more complex systems
+/// where the span information might be directly embedded within the span identifier itself.
+///
+/// ## Inspirations
+///
+/// This design is inspired by:
+/// - [rust-analyzer](https://github.com/rust-lang/rust-analyzer/blob/aa00ddcf654a35ba0eafe17247cf189958d33182/crates/span/src/lib.rs)
+/// - [rust](https://doc.rust-lang.org/stable/nightly-rustc/rustc_span/struct.Span.html)
+///
+/// ## Comparison
+///
+/// ### rust-analyzer
+///
+/// In rust-analyzer, spans are directly represented as `TextRange` during syntax parsing.
+/// In later stages, they are replaced with a more complex `Span` struct. Our `SpanId` is
+/// more compact, and the actual span data is looked up in storage.
+///
+/// ### rustc
+///
+/// rustc splits span information between `Span` and `SpanData`, where `Span` is more complex
+/// but also contains more intricate encoding. Our `SpanId` is simpler and references data
+/// stored elsewhere.
+///
+/// [`SpanStorage`]: self::storage::SpanStorage
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct SpanId(u32);
+
+impl SpanId {
+    pub(crate) const fn new(id: u32) -> Self {
+        Self(id)
+    }
+
+    pub(crate) const fn value(self) -> u32 {
+        self.0
+    }
+}
+
+/// Represents a full span in a file.
+///
+/// A span is a range of text within a file, along with a reference to the file itself, every span
+/// is relative to the parent span, if no parent is provided, the span is considered to be absolute.
+///
+/// Any span may have additional metadata attached, such as a JSON Pointer, for any JSON based
+/// frontends, this data is always optional, as some spans, for example for malformed code, may not
+/// have any additional data.
+pub trait Span {
+    /// Optional parent span, if any.
+    fn parent_id(&self) -> Option<SpanId>;
```

Comment:
You insert them one after the other, so first: `let parent_id = storage.insert(parent)` and then: `parent_id: Some(parent_id)`.

Reply:
So like:

```
A <- B <- C
```

without `B` and `C` already being in the store? Then yes that is correct. This keeps the implementation simple and ensures that you really do commit as soon as possible to the storage, so that you get an id, so that you can actually reference it.

You could instead also insert a `SpanNode`, but then you have the problem of trying to find duplicates, etc. etc. You circumvent that whole issue by inserting as soon as possible (and is also what e.g. `la_arena` or `ego_tree` and friends do)

Reply:
Let's see how this is going to look like when we actually use this. I have concerns that this will be annoying but let's see 🙂